src/core/servers.c

```diff
@@ -660,6 +660,14 @@ void server_change_nick(SERVER_REC *server, const char *nick)
 	signal_emit("server nick changed", 1, server);
 }
 
+void server_stash_meta(SERVER_REC *server, const char *meta_key, const char *meta_value)
+{
+	server->current_incoming_meta =
+	    g_slist_prepend(server->current_incoming_meta, g_strdup(meta_key));
+	server->current_incoming_meta =
+	    g_slist_prepend(server->current_incoming_meta, g_strdup(meta_value));
```

Comment:
Why not a hash table? It's not like linked lists are particularly efficient.

Reply:
they are space efficient and easy to push though. do you think a hash table has so much advantages that you want me to reimplement this as hash instead?

Reply:
note that this stash in particular is only pushed to, then it is iterated over and turned into a TEXT_BUFFER_META_REC

Reply:
>they are space efficient

I figured I'd test that claim, wrote [this test program][1] which puts a bunch of key/value pairs in gslist or ghashtable with random 8 char strings (so not testing for allocator efficiency here, which linked lists rely a lot more on)

Ran the thing with massif and uploaded [screenshots of massif-visualizer here][2]. You can view most of those side-by-side (except the 100 ones where the y axis is 90kb in one and 80kb in the other, annoying.)

In the graphs, the memory usage of linked lists increases linearly in a straight line, and hash tables are similarly linear but much more steppy, sometimes more memory efficient, sometimes less. The difference doesn't look significant. Area under curve looks more or less the same by eyeballing.

The peak memory usage:

|pairs|list|hash|
|------|----|----|
|10|22.2 KiB|22.3 KiB|
|100|26.5 KiB|26.0 KiB|
|1000|70.0 KiB|79.2 KiB|
|10000|506.8 KiB|514.6 KiB|

Can go either way. Mostly depends on which side of the curve you land on. Not really significant.

Also, I read this rant a while ago and it really stuck with me:

https://rust-unofficial.github.io/too-many-lists/#an-obligatory-public-service-announcement

It addreses a bunch of common points in favor of linked lists, but mostly from the perspective of using vectors instead (in our case, GArray). Except we're storing key-value data, and we already have a nice data structure for that.

>easy to push

That'd be `g_hash_table_insert(hash_table, meta_key, meta_value)` vs the four lines this comment thread highlighted. 


[1]: https://dump.dequis.org/indexed/massifs-hash-vs-list/asd.c
[2]: https://dump.dequis.org/indexed/massifs-hash-vs-list/graphs/

Reply:
interesting. you have to create the hash table first though